src/storage/gpu_swap_history.cc

```diff
@@ -236,18 +244,47 @@ void MemoryHistory::StartIteration() {
   for (int i = 0; i < NUMBER_OF_GPU; i++) {
     dev_history_[i].curr_idx = 0;
   }
-  if (iteration_idx_ <= 2 || swap_algorithm_ == "LRU") {
+  // LRU needs to record every iteration. As a result, it is mandatory to do LRU
+  // recording even at kBeginRecordAt iteration because the desired swapping
+  // algorithm will kick in from (kBeginRecordAt + 1) iteration.
+  if (iteration_idx_ <= kBeginRecordAt || swap_algorithm_ == "LRU") {
     pre_recording_ = true;
   }
-  if (iteration_idx_ == 2) {
+  if ((adaptive_history_ && iteration_idx_ >= kBeginRecordAt) ||
+      iteration_idx_ == kBeginRecordAt) {
+    // Each history is stored in a cyclic list (mimiced by std::list).
+    // The LAST history in a cyclic list is used to be recorded in current
+    // iteration. The (LAST-1) history is used by DoDecide(). However, if
+    // adaptive_history_ is false, The LAST history is used by DoDecide().
+    for (int i = 0; i < NUMBER_OF_GPU; i++) {
+      auto& history = dev_history_[i];
+      if (history.all_ordered_history.size() ==
+          DeviceHistory::kMaxPreservedIteration) {
+        history.all_ordered_history.splice(history.all_ordered_history.end(),
+                                           history.all_ordered_history,
+                                           history.all_ordered_history.begin());
+        history.all_handle_history.splice(history.all_handle_history.end(),
+                                          history.all_handle_history,
+                                          history.all_handle_history.begin());
+      } else {
+        size_t size = history.all_ordered_history.size();
+        history.all_ordered_history.resize(size + 1);
+        history.all_handle_history.resize(size + 1);
+      }
+      auto ordered_it = history.all_ordered_history.rbegin();
+      auto handle_it  = history.all_handle_history.rbegin();
+      if (history.all_ordered_history.size() > 1) {
+        ordered_it = std::next(ordered_it, 1);
+        handle_it = std::next(handle_it, 1);
+      }
+      history.ordered_history = &(*ordered_it);
```

Comment:
size == 1 can be either adaptive_history is true or iteration_idx_ == kBeginRecordAt. For size > 1, it can only be adaptive_history is true. So I'm not sure what did you mean. Can you explain in more detail?